Bug 4885: Excessive memory usage when running out of descriptors

craiggowing opened the pull request:
TcpAcceptor now stops listening when it cannot accept due to FD limits.
We also no longer defer/queue the same limited TcpAcceptor multiple
times. These changes prevent unbounded memory growth and improve
performance of Squids running out of file descriptors. They should have
no impact on other Squids.

craiggowing commented:
Thanks for the feedback @rousskov, I've attempted to address both points.

For the first point I have changed from using a vector to a set, so the listening socket will be added at most once. I don't think this is strictly necessary since from debugging the change for the second point it doesn't seem to ever get added when already present, but using a set will prevent memory issues it if it ever does happen.

For the second point I have prevented the call to "SetSelect" when deferred, so the socket is only polled next after the call to "Comm::AcceptLimiter::kick" made from "comm_close_complete" when a socket has closed.

I've found with the second change that the CPU usage is noticeably lower than with the first patch when sending more requests than the max FDs, I would speculate this is due to significantly less context switching taking place.

yadij commented on src/comm/AcceptLimiter.h:
```diff
@@ -56,7 +56,7 @@ class AcceptLimiter
     static AcceptLimiter Instance_;
 
     /** FIFO queue */
```
This comment needs to be updated. Set are not ordered so at the very least this cannot be a FIFO.

craiggowing replied:
I've removed the comment entirely. I was trying to think of something to replace it with, but "set of deferred TcpAcceptors" seemed a bit redundant.

yadij commented on src/comm/AcceptLimiter.cc:
```diff
@@ -26,37 +26,32 @@ Comm::AcceptLimiter::defer(const Comm::TcpAcceptor::Pointer &afd)
-            ++abandonedClients;
-        }
+    std::set<TcpAcceptor::Pointer>::iterator it;
+    it = deferred_.find(afd);
```
The 'it' local can have type const auto if it works, otherwise just auto and avoid the C-style pre-define.

craiggowing replied:
Changed both to "const auto"

craiggowing commented on src/comm/AcceptLimiter.cc:
```diff
@@ -26,37 +26,32 @@ Comm::AcceptLimiter::defer(const Comm::TcpAcceptor::Pointer &afd)
 {
     ++ (afd->isLimited);
```
Removed isLimited since it has no real purpose now, you can check if it is limited by looking to see if it exists in the set.

yadij replied:
I was just thinking that this member could be used as a flag now to avoid the search when adding. Does not matter enough to add back though.

rousskov replied:
There should be no search when adding because the code should be written so that we never add twice (roughly speaking). If maintaining a boolean flag is the best way to maintain that invariant, we can convert this counter into such a flag, but I hope there are better ways to accomplish the same thing.

yadij commented:
Updated the comment and title to resolve description complaints from Anubis.

I am marking this as approved since my issues are resolved.

craiggowing commented:
Updated again - I have reverted some of the changed required to use a set and changed to a deque as suggested.

I tested set vs deque with multiple listening sockets and as suspected when using a set requests where only handled by one listening socket if that has more connections than FDs, the other connections just stalled. Once using a deque everything balanced nicely across all listening sockets. - however since we are only adding each listening at most once to the deque we will not get any weighting, it will always be split evenly.

I'm not sure there would be an easy way of weighting this based on the backlogs themselves, even if we were adding the listener multiple times to the deque, as long as at least one thing is in the backlog it will be added once each time we get the SetSelect callback for every listener, which would give the same even split of traffic while using a lot more memory.

yadij commented:
IMO that balancing style is better than weighted. It will make DoS attempts have to flood every receiving port of a proxy in order to block service completely, instead of only having to locate and flood one.

One very useful implication of that is the possibility of having a management port which remains somewhat responsive at times when the main traffic port is overloading.

rousskov commented:
Round-robin is the right default because it is simple and works "well enough" for most use cases. No need to complicate things until somebody needs a different load balancing policy.

rousskov commented on src/comm/AcceptLimiter.cc:
```diff
@@ -24,20 +24,17 @@ Comm::AcceptLimiter::Instance()
-            debugs(5, 5, afd->conn << " x" << afd->isLimited);
+    for (auto it = deferred_.begin(); it != deferred_.end(); ++it) {
+        if (*it == afd) {
+            *it = NULL; // fast. kick() will skip empty entries later.
```
s/NULL/nullptr/ because you have to change this code line for other reasons

rousskov commented on src/comm/AcceptLimiter.h:
```diff
@@ -11,13 +11,13 @@
 
 /**
- * FIFO Queue holding listener socket handlers which have been activated
+ * Deque holding listener socket handlers which have been activated
```
Let's restore `FIFO` to document the order. You do not have to modify this comment line at all AFAICT: The words "FIFO queue" are a high-level description of how deferred listeners are stored/ordered. The specific C++ type used to implement that storage does not matter here.

rousskov commented on src/comm/TcpAcceptor.cc:
```diff
@@ -318,6 +315,7 @@ Comm::TcpAcceptor::acceptNext()
     Must(IsConnOpen(conn));
     debugs(5, 2, HERE << "connection on " << conn);
     acceptOne();
+    SetSelect(conn->fd, COMM_SELECT_READ, doAccept, this, 0);
```
Please move this SetSelect() call to the end of the acceptOne() method to

* avoid calling SetSelect() twice on Comm::NOMESSAGE and
* avoid calling SetSelect() at all after fatal accepting failures.

rousskov commented on src/comm/AcceptLimiter.cc:
```diff
@@ -24,20 +24,17 @@ Comm::AcceptLimiter::Instance()
 {
-    ++ (afd->isLimited);
-    debugs(5, 5, afd->conn << " x" << afd->isLimited);
+    debugs(5, 5, "deferring " << afd->conn);
```
Instead of printing `deferring` (which is redundant because Squid will print `defer` (i.e. the method name) automatically), please report the current queue size. For example:

```
debugs(5, 5, afd->conn << "; already queued: " << deferred_.size());
```

rousskov commented on src/comm/AcceptLimiter.cc:
```diff
@@ -24,20 +24,17 @@ Comm::AcceptLimiter::Instance()
+    for (auto it = deferred_.begin(); it != deferred_.end(); ++it) {
+        if (*it == afd) {
+            *it = NULL; // fast. kick() will skip empty entries later.
             ++abandonedClients;
```
Since we never defer two pointers to the same TcpAcceptor object, we should debug() and return here. And there will be no need to count the number of matches after that change either -- it will always be one here and zero after the loop. The debug message after the loop should then be adjusted to report the failure to find `afd->conn` in `deferred_.size()`-element queue.

rousskov commented:
> when connections reach max FD

The problems we are addressing start earlier, when the number of open descriptors exceeds (Squid_MaxFD-RESERVED_FD), rather than Squid_MaxFD. I suggest changing the title to something like `Bug 4885: Excessive memory usage when running out of file descriptors`.

Besides reviewing that suggestion, please also check the adjusted PR [description](https://github.com/squid-cache/squid/pull/291#issue-217616246) for accuracy. Whatever we put there will become the official commit message body.

craiggowing commented:
Updated the title, description looks good to me.
